File: Content/Algorithms/Sorting/Python/Radix.py
def CountingSort(arr, place):
    temp = arr
    size = len(temp)
    res = [0] * len(temp)
    count = [0] * 10

    for i in range(0, len(temp)):
        index = arr[i] // place
        count[index % 10] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    i = size - 1
    while i >= 0:
        index = arr[i] // place
        res[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1


    for i in range(0, size):
        temp[i] = res[i]

    return temp


def RadixSort( arr ):
    sortedArr = arr
    max_element = max( sortedArr )

    place = 1
    while max_element // place > 0:
        CountingSort( sortedArr , place )
        place *= 10

    return sortedArr

if __name__ == "__main__":
    from random import randint
    from time import time

    # running test cases:
    testArr = [ randint(0, 1000) for i in range(50) ]

    print("Unsorted Array: ", testArr, "\n"); start = time()
    print("Sorted Array: ", RadixSort(testArr), "\n")
    print("Time Taken: ", time() - start)

File: Content/Sorting/Golang/sorting/Radix.go
package sorting

//RadixSort sorts an array using the ***radix sort algorithm***
func RadixSort(arr *[]int64) {

}

/*
RadixSorted sorts an array using the ***radix sort algorithm*** and returns the sorted array

! Note: This function doesn't **modify** the array passed into the function
*/
func RadixSorted(arr []int64) []int64 {
	res := arr
	maxElem := _max(res)

	place := int64(1)
	for int(maxElem/place) > 0 { //needs work ...
		_countingSort(res, place)
		place *= 10
	}

	return res
}

func _countingSort(arr []int64, place int64) []int64 {
	temp := arr
	size := len(temp)
	res := make([]int64, size)
	count := make([]int64, 10)

	for i := 0; i < size; i++ {
		index := int64(arr[i] / place)
		count[index%10]++
	}
	for i := 1; i < 10; i++ {
		count[i] += count[i-1]
	}
	i := size - 1
	for i >= 0 {
		index := int64(arr[i] / place)
		res[count[index%10]-1] = arr[i]
		count[index%10]--
		i--
	}

	return arr
}
